### PR TITLE
Automate updating out of date provider ids on application choices

### PR DESCRIPTION
**Pull request by cwrw**
## Context

We currently store denormalised course data on application choices. One of those fields holds all provider ids that the course option and current course option relate to (provider and accredited provider).

This field is key to show the application to the relevant providers.
We currently pull course data from publish, which is updating accredited providers regularly. When we sync with the publish API we are updating the course accredited provider, but not application choices. This means that application choices will have stale provider ids

## Changes proposed in this pull request

- Add worker to update stale provider ids on application choices
- Run this job every hour

## Guidance to review
Did I miss anything?
Needed to do some test rejigging to appease Rubocop overlord

## Link to Trello card

https://trello.com/c/Q5ovlGxk/327-automate-accredited-provider-updates-to-courses

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Review by tomas-stefano — APPROVED** [review]
🚀

File: app/workers/update_out_of_date_provider_ids_on_application_choices.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  def perform
+    return unless out_of_date_application_choices.any?
+
+    update_out_of_date_application_choices
+  end
+
+private
+
+  def out_of_date_application_choices
+    @out_of_date_application_choices ||= FindApplicationChoicesWithOutOfDateProviderIds.call
```
Comment (carlosmartinez) [inline app/workers/update_out_of_date_provider_ids_on_application_choices.rb:13]: what do you think of the convention of using an underscore for these memoized variables? as in `@_out_of_date_application_choices`. i quite like it.

**Review by carlosmartinez — APPROVED** [review]
great. couple of comments/suggestions which you're free to follow or ignore, but i'm happy if it's merged as-is.

File: app/workers/update_out_of_date_provider_ids_on_application_choices.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+class UpdateOutOfDateProviderIdsOnApplicationChoices
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/workers/update_out_of_date_provider_ids_on_application_choices.rb:1]: With the existing query service name `FindApplicationChoicesWithOutOfDateProviderIds` my preference is to keep this name similar as well. I feel it also expresses explicitly what this class does

File: app/workers/update_out_of_date_provider_ids_on_application_choices.rb (hunk 1 above)
Comment (cwrw) [reply app/workers/update_out_of_date_provider_ids_on_application_choices.rb]: done

**Pull request merged**